compiler/rustc_span/src/source_map.rs

```diff
@@ -859,14 +859,15 @@ impl SourceMap {
         }
         let start_of_next_point = sp.hi().0;
 
-        let width = self.find_width_of_character_at_span(sp.shrink_to_hi(), true);
+        let width = self.find_width_of_character_at_span(sp, true);
+        debug_assert!(width > 0);
         // If the width is 1, then the next span should point to the same `lo` and `hi`. However,
         // in the case of a multibyte character, where the width != 1, the next span should
         // span multiple bytes to include the whole character.
         let end_of_next_point =
-            start_of_next_point.checked_add(width - 1).unwrap_or(start_of_next_point);
+            start_of_next_point.checked_add(width).unwrap_or(start_of_next_point);
```

Comment:
I've always found this bit confusing, but I'm wondering if this comment is now out-of-date?

The `next_point()` function used to return an empty span (when the "width is 1" as in the comment) pointing just before the next character. This change seems to change it so that it is a non-empty span containing the next character.

I'm also wondering if the name of the function is also a bit misleading? The original `next_point` function always returned an empty span (a "point", not a region). This was changed in #47420 to have the somewhat confusing behavior where an empty span would move forward to span the next character (non-empty), and a non-empty span would return an empty span pointing just past the end. It's not clear from that PR why that behavior was chosen.

I also think this function is still somewhat buggy with respect to multi-byte characters. Below is a unittest which illustrates some oddities. Perhaps it might be good to add this unittest?

```rust
#[test]
fn next_point() {
    let sm = SourceMap::new(FilePathMapping::empty());
    sm.new_source_file(PathBuf::from("example.rs").into(), "a…b".to_string());
    let span = Span::with_root_ctxt(BytePos(0), BytePos(0));
    // This confusingly does not advance the span?
    let span = sm.next_point(span);
    assert_eq!(span.lo().0, 0);
    assert_eq!(span.hi().0, 0); // ERROR: This should probably be 1?

    let span = Span::with_root_ctxt(BytePos(0), BytePos(1));
    let span = sm.next_point(span);
    assert_eq!(span.lo().0, 1);
    assert_eq!(span.hi().0, 4);

    // This creates an invalid span, slicing in the middle of a multi-byte char?
    let span = Span::with_root_ctxt(BytePos(1), BytePos(1));
    let span = sm.next_point(span);
    assert_eq!(span.lo().0, 1);
    assert_eq!(span.hi().0, 2); // ERROR: This should probably be 4?

    let span = Span::with_root_ctxt(BytePos(1), BytePos(4));
    let span = sm.next_point(span);
    assert_eq!(span.lo().0, 4);
    assert_eq!(span.hi().0, 5);
    // This creates an invalid span, pointing past the end of the file?
    let span = sm.next_point(span);
    assert_eq!(span.lo().0, 5);
    assert_eq!(span.hi().0, 6); // ERROR: This should probably be 5?

    // Empty span pointing just past the last byte.
    let span = Span::with_root_ctxt(BytePos(5), BytePos(5));
    let span = sm.next_point(span);
    assert_eq!(span.lo().0, 5);
    assert_eq!(span.hi().0, 6); // ERROR: This should probably be 5?
}
```

Reply:
This unit test implied another issue:

```rust
// This creates an invalid span, slicing in the middle of a multi-byte char?
    let span = Span::with_root_ctxt(BytePos(1), BytePos(1));
    let span = sm.next_point(span);
    assert_eq!(span.lo().0, 1);
    assert_eq!(span.hi().0, 2);
```
When `lo` and `hi` are same, we just reutrn the width with 1:
https://github.com/rust-lang/rust/blob/194140bef501ad3acb00d57c20fb80ee34aa1d3b/compiler/rustc_span/src/source_map.rs#L877

So the result is `span.hi().0` will be 2.

Reply:
Thanks @ehuss,

I fixed all other issues in the test cases, except for this one:

```rust
 let span = sm.next_point(span);
    assert_eq!(span.lo().0, 0);
    assert_eq!(span.hi().0, 0); // ERROR: This should probably be 1?
```

It return the same span because it's a dummy span, something special for macros..